umpire +rocm build fails: CMake Error in examples/CMakeLists.txt: HIP_ARCHITECTURES is empty for target

### Steps to reproduce the issue

`umpire@6.0.0 +rocm amdgpu_target=gfx908` fails to build using:
* `spack@develop` (f81d84dfc6b1c3a821dffab785e62ecd79f130ab from `Tue Dec 7 12:30:14 2021 +0100`
* Ubuntu 20.04, x86_64
* GCC 9.3.0
* ROCm 4.3.1, defined externally, installed via apt

Concrete spec: [umpire.spec.json.txt](https://github.com/spack/spack/files/7681158/umpire.spec.json.txt)

```
Concretized
--------------------------------
umpire@6.0.0%gcc@9.3.0+c~cuda~deviceconst+examples~fortran~ipo~numa~openmp+rocm+shared amdgpu_target=gfx908 build_type=RelWithDebInfo tests=none arch=linux-ubuntu20.04-x86_64
    ^blt@0.4.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^cmake@3.21.4%gcc@9.3.0~doc+ncurses+openssl+ownlibs~qt build_type=Release arch=linux-ubuntu20.04-x86_64
            ^ncurses@6.2%gcc@9.3.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-x86_64
                ^pkgconf@1.8.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^openssl@1.1.1l%gcc@9.3.0~docs certs=system arch=linux-ubuntu20.04-x86_64
                ^perl@5.34.0%gcc@9.3.0+cpanm+shared+threads arch=linux-ubuntu20.04-x86_64
                    ^berkeley-db@18.1.40%gcc@9.3.0+cxx~docs+stl patches=b231fcc4d5cff05e5c3a4814f6a5af0e9a966428dc2176540d2c05aff41de522 arch=linux-ubuntu20.04-x86_64
                    ^bzip2@1.0.8%gcc@9.3.0~debug~pic+shared arch=linux-ubuntu20.04-x86_64
                        ^diffutils@3.8%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^libiconv@1.16%gcc@9.3.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
                    ^gdbm@1.19%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^readline@8.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^zlib@1.2.11%gcc@9.3.0+optimize+pic+shared arch=linux-ubuntu20.04-x86_64
    ^camp@0.2.2%gcc@9.3.0~cuda~ipo+rocm~tests amdgpu_target=gfx908 build_type=RelWithDebInfo arch=linux-ubuntu20.04-x86_64
        ^hip@4.3.1%gcc@9.3.0~ipo build_type=Release patches=2a4190477b7d9206b9cd8d70770ba0bc007273cbe54772efb12f9ca2e37c0392,3d20c8737ff8d0a00053f6bf23009df060fb26d39545aa34c74428b69d02257a,99190b4616edb362d48f9b265c3018a3c6339481b0729d9fe46185fca25bc54b,e276c4acf3d37712b6bea306fea34f539d3c4f743471e9da208b5eb17b16ae67 arch=linux-ubuntu20.04-x86_64
            ^comgr@4.3.1%gcc@9.3.0~ipo build_type=Release arch=linux-ubuntu20.04-x86_64
                ^llvm-amdgpu@4.3.1%gcc@9.3.0~ipo+openmp+rocm-device-libs build_type=Release patches=d999f3b235e655ee07f6dd2590302082feaa06d32c5c6b53aae9c5cf1e45b644 arch=linux-ubuntu20.04-x86_64
                    ^hwloc@2.6.0%gcc@9.3.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl+pci~rocm+shared arch=linux-ubuntu20.04-x86_64
                        ^libpciaccess@0.16%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^libtool@2.4.6%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                                ^m4@1.4.19%gcc@9.3.0+sigsegv patches=9dc5fbd0d5cb1037ab1e6d0ecc74a30df218d0a94bdd5a02759a97f62daca573,bfdffa7c2eb01021d5849b36972c069693654ad826c1a20b53534009a4ec7a89 arch=linux-ubuntu20.04-x86_64
                                    ^libsigsegv@2.13%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^util-macros@1.19.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^libxml2@2.9.12%gcc@9.3.0~python arch=linux-ubuntu20.04-x86_64
                            ^xz@5.2.5%gcc@9.3.0~pic libs=shared,static arch=linux-ubuntu20.04-x86_64
                    ^libelf@0.8.13%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^perl-data-dumper@2.173%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^python@3.8.12%gcc@9.3.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87,4c2457325f2b608b1b6a2c63087df8c26e07db3e3d493caf36a56f0ecf6fb768,f2fd060afc4b4618fe8104c4c5d771f36dc55b1db5a4623785a4ea707ec72fb4 arch=linux-ubuntu20.04-x86_64
                        ^expat@2.4.1%gcc@9.3.0+libbsd arch=linux-ubuntu20.04-x86_64
                            ^libbsd@0.11.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                                ^libmd@1.0.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^gettext@0.21%gcc@9.3.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu20.04-x86_64
                            ^tar@1.34%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^libffi@3.3%gcc@9.3.0 patches=26f26c6f29a7ce9bf370ad3ab2610f99365b4bdd7b82e7c31df41a3370d685c0 arch=linux-ubuntu20.04-x86_64
                        ^sqlite@3.36.0%gcc@9.3.0+column_metadata+fts~functions~rtree arch=linux-ubuntu20.04-x86_64
                        ^util-linux-uuid@2.36.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^z3@4.8.9%gcc@9.3.0~gmp~ipo~python build_type=RelWithDebInfo arch=linux-ubuntu20.04-x86_64
                ^rocm-cmake@4.3.1%gcc@9.3.0~ipo+ldconfig build_type=Release arch=linux-ubuntu20.04-x86_64
            ^hip-rocclr@4.3.1%gcc@9.3.0~ipo build_type=Release arch=linux-ubuntu20.04-x86_64
                ^hsa-rocr-dev@4.3.1%gcc@9.3.0+image~ipo+shared build_type=Release patches=71e6851d9be8113bfb8d71b66a3171e0d0401bae5e6f161c9e7fe32558261a46 arch=linux-ubuntu20.04-x86_64
                    ^hsakmt-roct@4.3.1%gcc@9.3.0~ipo+shared build_type=Release arch=linux-ubuntu20.04-x86_64
                        ^numactl@2.0.14%gcc@9.3.0 patches=4e1d78cbbb85de625bad28705e748856033eaafab92a66dffd383a3d7e00cc94,62fc8a8bf7665a60e8f4c93ebbd535647cebf74198f7afafec4c085a8825c006,ff37630df599cfabf0740518b91ec8daaf18e8f288b19adaae5364dc1f6b2296 arch=linux-ubuntu20.04-x86_64
                            ^autoconf@2.69%gcc@9.3.0 patches=35c449281546376449766f92d49fc121ca50e330e60fefcfc9be2af3253082c2,7793209b33013dc0f81208718c68440c5aae80e7a1c4b8d336e382525af791a7,a49dd5bac3b62daa0ff688ab4d508d71dbd2f4f8d7e2a02321926346161bf3ee arch=linux-ubuntu20.04-x86_64
                            ^automake@1.16.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^xxd-standalone@8.2.1201%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^mesa18@18.3.6%gcc@9.3.0+glx~llvm~opengles+osmesa patches=55a5611ca9fcbe8324c4d68a07b338134954ff12c5b122dc78ff376f012a1414 swr=none arch=linux-ubuntu20.04-x86_64
                    ^binutils@2.37%gcc@9.3.0~gas~gold~headers~interwork~ld~libiberty~lto+nls+plugins libs=shared,static arch=linux-ubuntu20.04-x86_64
                    ^bison@3.8.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^flex@2.6.3%gcc@9.3.0+lex~nls arch=linux-ubuntu20.04-x86_64
                        ^findutils@4.8.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^glproto@1.4.17%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^libx11@1.7.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^inputproto@2.3.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^kbproto@1.0.7%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^libxcb@1.14%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^libpthread-stubs@0.4%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^libxau@1.0.8%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                                ^xproto@7.0.31%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^libxdmcp@1.1.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^xcb-proto@1.14.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^xextproto@7.3.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^xtrans@1.3.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^libxext@1.3.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^py-mako@1.1.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^py-markupsafe@2.0.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^py-setuptools@59.4.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^rocminfo@4.3.1%gcc@9.3.0~ipo build_type=Release arch=linux-ubuntu20.04-x86_64
            ^roctracer-dev-api@4.3.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
```

```
$> spack install -f ./umpire.spec.json
...
==> Installing umpire-6.0.0-z7qnz373acjrzch7bqtyh3gcrzyrck4b
==> No binary for umpire-6.0.0-z7qnz373acjrzch7bqtyh3gcrzyrck4b found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/git//LLNL/Umpire.git/v6.0.0.tar.gz
==> Warning: Fetching from mirror without a checksum!
  This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
==> No patches needed for umpire
==> umpire: Executing phase: 'initconfig'
==> umpire: Executing phase: 'cmake'
==> Error: ProcessError: Command exited with status 1:
    'cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/umpire-6.0.0-z7qnz373acjrzch7bqtyh3gcrzyrck4b' '-DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=OFF' '-DCMAKE_INSTALL_RPATH:STRING=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/umpire-6.0.0-z7qnz373acjrzch7bqtyh3gcrzyrck4b/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/umpire-6.0.0-z7qnz373acjrzch7bqtyh3gcrzyrck4b/lib64;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/camp-0.2.2-6etnaqofpmhptmtqs3s4mezobhm4ktmr/lib;/opt/rocm-4.3.1/hip/lib;/opt/rocm-4.3.1/lib;/opt/rocm-4.3.1/llvm/lib' '-DCMAKE_PREFIX_PATH:STRING=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/camp-0.2.2-6etnaqofpmhptmtqs3s4mezobhm4ktmr;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/blt-0.4.1-hw4qld44gjpoigfsarp3m7fyvfee3ezx;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/cmake-3.21.4-55vqjfj2bk55ga4fj3erj5tpzugh3t3b;/opt/rocm-4.3.1/llvm;/opt/rocm-4.3.1/;/opt/rocm-4.3.1/hip' '-C' '/tmp/root/spack-stage/spack-stage-umpire-6.0.0-z7qnz373acjrzch7bqtyh3gcrzyrck4b/spack-src/1045365837ee-linux-ubuntu20.04-x86_64-gcc@9.3.0.cmake' '/tmp/root/spack-stage/spack-stage-umpire-6.0.0-z7qnz373acjrzch7bqtyh3gcrzyrck4b/spack-src'

37 errors found in build log:
     75     -- Checking for std::filesystem
     76     -- Performing Test UMPIRE_ENABLE_FILESYSTEM
     77     -- Performing Test UMPIRE_ENABLE_FILESYSTEM - Failed
     78     -- std::filesystem NOT found, using POSIX
     79     -- Host Shared Memory Disabled
     80     -- Configuring done
  >> 81     CMake Error in examples/CMakeLists.txt:
     82       HIP_ARCHITECTURES is empty for target "alloc_access".
     83
     84
  >> 85     CMake Error in examples/CMakeLists.txt:
     86       HIP_ARCHITECTURES is empty for target "allocator".
     87
     88
  >> 89     CMake Error in examples/CMakeLists.txt:
     90       HIP_ARCHITECTURES is empty for target "vector_allocator".
     91
     92
  >> 93     CMake Error in examples/CMakeLists.txt:
     94       HIP_ARCHITECTURES is empty for target "size_limiter".
...
```



### Information on your system

* **Spack:** 0.17.0-446-3db39db0b3
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-broadwell
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/7681163/spack-build-out.txt)
[spack-build-02-cmake-out.txt](https://github.com/spack/spack/files/7681168/spack-build-02-cmake-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/7681169/spack-build-env.txt)

@davidbeckingsale 
@white238

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/umpire/package.py
+++ b/var/spack/repos/builtin/packages/umpire/package.py
@@ -73,6 +73,7 @@ class Umpire(CachedCMakePackage, CudaPackage, ROCmPackage):
 
     depends_on('cmake@3.8:', type='build')
     depends_on('cmake@3.9:', when='+cuda', type='build')
+    depends_on('cmake@:3.20', when='+rocm', type='build')
 
     depends_on('blt@0.4.1:', type='build', when='@6.0.0:')
     depends_on('blt@0.4.0:', type='build', when='@4.1.3:5.0.1')

